### PR TITLE
[Bug](predicate) fix wrong result of AcceptNullPredicate

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
## Proposed changes
fix wrong result of AcceptNullPredicate

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
run buildall

**Review by HappenLee — COMMENTED** [review]
PLEASE ADD TEST

File: be/src/olap/accept_null_predicate.h [hunk 1]
```diff
@@ -158,21 +156,26 @@ class AcceptNullPredicate : public ColumnPredicate {
             }
             // create selected_flags
             uint16_t max_idx = sel[size - 1];
+            std::vector<uint16_t> old_sel(size);
```
Comment (yiguolei) [inline be/src/olap/accept_null_predicate.h:159]: please add some test

File: be/src/olap/accept_null_predicate.h [hunk 2]
```diff
@@ -158,21 +156,26 @@ class AcceptNullPredicate : public ColumnPredicate {
...
-                memcpy(selected.data(), nullable_col.get_null_map_data().data(),
-                       (max_idx + 1) * sizeof(bool));
+                std::vector<uint8_t> selected(max_idx + 1, 0);
+                const auto* nullmap = nullable_col.get_null_map_data().data();
                 // add rows selected by _nested->evaluate
                 for (uint16_t i = 0; i < new_size; ++i) {
                     uint16_t row_idx = sel[i];
                     selected[row_idx] = true;
                 }
+                for (uint16_t i = 0; i < size; ++i) {
```
Comment (yiguolei) [inline be/src/olap/accept_null_predicate.h:175]: add comment here

File: be/src/olap/accept_null_predicate.h [hunk 3]
```diff
@@ -158,21 +156,26 @@ class AcceptNullPredicate : public ColumnPredicate {
             }
             // create selected_flags
             uint16_t max_idx = sel[size - 1];
+            std::vector<uint16_t> old_sel(size);
+            memcpy(old_sel.data(), sel, sizeof(uint16_t) * size);
 
             const auto& nullable_col = assert_cast<const vectorized::ColumnNullable&>(column);
             // call nested predicate evaluate
             uint16_t new_size = _nested->evaluate(nullable_col.get_nested_column(), sel, size);
```
Comment (yiguolei) [inline be/src/olap/accept_null_predicate.h:165]: sel  1,2,3,7,8,9

sel 1,2,3,4,5,6,7,8,9

**Comment by BiteTheDDDDt** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

File: be/src/olap/accept_null_predicate.h [hunk 4]
```diff
@@ -64,11 +63,14 @@ class AcceptNullPredicate : public ColumnPredicate {
...
         if (column.has_null()) {
+            std::vector<uint8_t> original_flags(size);
+            memcpy(original_flags.data(), flags, size);
+
             const auto& nullable_col = assert_cast<const vectorized::ColumnNullable&>(column);
             _nested->evaluate_and(nullable_col.get_nested_column(), sel, size, flags);
             const auto& nullmap = nullable_col.get_null_map_data();
             for (uint16_t i = 0; i < size; ++i) {
-                flags[i] |= nullmap[sel[i]];
+                flags[i] |= (original_flags[i] && nullmap[sel[i]]);
```
Comment (yiguolei) [inline be/src/olap/accept_null_predicate.h:73]: 我觉得这里不对吧
是不是应该是 flags[sel[i]] |= (original_flags[sel[i]] && nullmap[sel[i]]);

File: be/src/olap/accept_null_predicate.h [hunk 5]
```diff
@@ -77,20 +79,7 @@ class AcceptNullPredicate : public ColumnPredicate {
...
-            for (uint16_t i = 0; i < size; ++i) {
-                uint16_t idx = sel[i];
-                if (!flags[i] && nullable_col.is_null_at(idx)) {
-                    flags[i] = true;
-                }
-            }
-        } else {
-            _nested->evaluate_or(column, sel, size, flags);
-        }
+        DCHECK(false) << "should not reach here";
```
Comment (yiguolei) [inline be/src/olap/accept_null_predicate.h:82]: 直接用check，保证线上数据别出错，或者throw exception

**Review by yiguolei — APPROVED** [review]

**Comment by BiteTheDDDDt** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by xiaokang — APPROVED** [review]
LGTM

**Pull request merged**